morpheus/stages/doca/doca_source_stage.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2021-2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import pathlib
+import typing
+
+import mrc
+import typing_utils
+from mrc.core import operators as ops
+
+from morpheus.cli import register_stage
+from morpheus.config import Config
+from morpheus.config import PipelineModes
+from morpheus.io.deserializers import read_file_to_df
+from morpheus.messages import MessageMeta
+from morpheus.pipeline.preallocator_mixin import PreallocatorMixin
+from morpheus.pipeline.single_output_source import SingleOutputSource
+from morpheus.pipeline.stream_pair import StreamPair
+
+logger = logging.getLogger(__name__)
+
+
+@register_stage("from-doca", modes=[PipelineModes.NLP], ignore_args=["cudf_kwargs"])
+class DocaSourceStage(PreallocatorMixin, SingleOutputSource):
+    """
+    Load messages from a file.
+
+    Source stage is used to load messages from a file and dumping the contents into the pipeline immediately. Useful for
+    testing performance and accuracy of a pipeline.
+
+    Parameters
+    ----------
+    c : `morpheus.config.Config`
+        Pipeline configuration instance.
+    filename : pathlib.Path, exists = True, dir_okay = False
+        Name of the file from which the messages will be read.
+    iterative : boolean, default = False, is_flag = True
+        Iterative mode will emit dataframes one at a time. Otherwise a list of dataframes is emitted. Iterative mode is
+        good for interleaving source stages.
+    repeat : int, default = 1, min = 1
+        Repeats the input dataset multiple times. Useful to extend small datasets for debugging.
+    filter_null : bool, default = True
+        Whether or not to filter rows with null 'data' column. Null values in the 'data' column can cause issues down
+        the line with processing. Setting this to True is recommended.
+    cudf_kwargs : dict, default = None
+        keyword args passed to underlying cuDF I/O function. See the cuDF documentation for `cudf.read_csv()` and
+        `cudf.read_json()` for the available options. With `file_type` == 'json', this defaults to ``{ "lines": True }``
+        and with `file_type` == 'csv', this defaults to ``{}``.
+    """
+
+    def __init__(
+        self,
+        c: Config,
+        nic_pci_address: str,
+        gpu_pci_address: str,
+    ):
+
+        super().__init__(c)
```

Comment:
We probably need some way of detecting if DOCA was built or not and throwing an error here

Reply:
Eventually we'll want to do something like that for optional modules by generating a python file (or copying this file to the appropriate location) on build, but we'd want to give more thought to that than we have time for now.